### DISCUSSION TITLE
[General] why sending 2 transactions to same contract with same method and very similar input data within minute end up consuming very different gas usage?

### DISCUSSION THREAD

**Discussion by cwhcheng** [zkSync-Community-Hub/zksync-developers#287] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I deployed a chainlink OffchainAggregator contract on zksync sepolia. 
- source code: https://github.com/smartcontractkit/libocr/blob/master/contract/OffchainAggregator.sol
- contract address: https://sepolia.explorer.zksync.io/address/0xEB5E7A57E117ac0367C9dc89E39485F8fFc497D3

I try to call `function transmit(bytes calldata _report, bytes32[] calldata _rs, bytes32[] calldata _ss, bytes32 _rawVs)` to update price data. I try to send out multiple transactions of same method with very similar input data to the same contract within minute but with different gas limit. Supposingly, different gas limit should not have big impact on actual gas usage given all those transactions performing very similar computation and data storage tasks. Yet, it ends up the one with more gas limit consuming 75% more gas than the one with less gas limit.

### What is the specific issue or error you're encountering?

gas usage is impacted significant by different gas limit

### Can you share the error messages or logs you're receiving, if any?

gas limit: 2,500,000        gas usage: ~2,000,000
- https://sepolia.explorer.zksync.io/tx/0x8451fec07ede5a5e4a4dd6104afdf03ceab2db555894a60ae93d7ecc4296c748 (24 Jan 15:37 batch#5721)
- https://sepolia.explorer.zksync.io/tx/0xff9a9ac9bca66710e845fdc630fe574405ce2e5e3e12bfb645781771681fe118 (24 Jan 11:57 batch#5717)

gas limit: 6,000,000    gas usage: ~3,500,000
- https://sepolia.explorer.zksync.io/tx/0xdd9d6a55698d45844263e75008ecd703fd66d893d24e3009286818c4f1fd1e00 (24 Jan 15:37 batch#5721)
- https://sepolia.explorer.zksync.io/tx/0x72c36d99a980e13098d853eb9ab6dfb3bd80e72d480ad09ba17e008208c0b2f1 (24 Jan 15:38 batch#5721)
- https://sepolia.explorer.zksync.io/tx/0x9c9b1751a02afb9c1b5ca422110e275bcb92406b2ee461a385981c3aa3594cfb (24 Jan 15:38 batch#5721)
- https://sepolia.explorer.zksync.io/tx/0xebda32d64f6dc918a5106f607f58db940d29089fb693ba54c1336a20047e02c2 (24 Jan 11:57 batch#5717)
- https://sepolia.explorer.zksync.io/tx/0xc1e3f92a01f1302a0c61f6463ffe5f08b877dac58cc1bba33f71db6157778eb2 (24 Jan 11:56 batch#5717)
- https://sepolia.explorer.zksync.io/tx/0xd49325aba8ffe606ca247f88e23ad9168f96d6caf01981ae25ba5d9c9e8d2393 (24 Jan 11:58 batch#5717)

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

https://github.com/smartcontractkit/libocr/blob/master/contract/OffchainAggregator.sol#L576

### Have you tried to isolate the problem, and if so, what were the results?

n/a

### What steps have you already taken to try to resolve the issue?

n/a

### Additional Details

thanks in advance!

**Comment by StanislavBreadless** [comment, marked as answer]
Hey, this is a known issue and it will be mostly fixed in the next upgrade.

The current behavior is made by the fact that we pre-charge an amount proportional to the `tx.gasLimit` to take into account that if the transaction is malicious, it might force us to close batch earlier than usual. Basically this is a security measure to ensure that if someone sends a transaction that is hard to aggregate with other transactions in the batch (for instance it consumed the entire pubdata limit and so no other transaction could be inserted into the batch), the sender of such txs would compensate for it. 

In the next release (very soon), we'll change how this protection works and we will no longer pre-charge and the users will pay only for the % of the batch they consumed. However, a small difference when different gasLimits are used will still be seen for a different reason:

- We employ state diff compression to minimize the costs for users. The rule of thumb there "the smaller the change is, the smaller amount of pubdata will be published with such a state diff". For instance, to compress that the nonce has been updated by 1, we just need to provide 2 bytes: 1 byte is for the id of the "Add" operation to denote that it has increased and the other one which contains the number 1. 
- When a transaction is started, the `tx.gasLimit * tx.maxFeePerGas` is pre-charged from your balance (note, the larger `gasLimit`, the larger difference in the value). At the time when we give out refunds, we just treat it as a larger state diff to be charged for. 

This difference will be much-much less noticeable that what you observe on the current mainnet.

**Reply by cwhcheng** [reply]
thanks for the answer!

**Comment by crimuntilve** [comment]
It seems like the gas usage discrepancy you're experiencing might be related to how gas estimation is performed in zkSync. Gas estimation can sometimes be imprecise, especially when dealing with complex transactions or interactions with external contracts.

To investigate further, you could try analyzing the bytecode of the contract and the transactions to see if there are any differences that might account for the gas usage discrepancy. Additionally, reaching out to the zkSync team or community for assistance and further insights might be helpful in resolving this issue.

**Comment by sayloonkenneth** [comment]
Issue: Gas usage significantly impacted by different gas limits when calling the transmit function of Chainlink OffchainAggregator contract on zkSync sepolia testnet.
To address this issue effectively, consider analyzing the contract's implementation to identify any gas-intensive operations or potential optimizations. Additionally, ensure that the gas limits set for transactions align with the actual computational requirements of the function being called.

**Comment by cbaca9** [comment]
Gas usage discrepancies can occur due to factors like storage changes or computational variations. 
Check the contract's logic for potential differences in gas consumption based on input data or storage state.
